nbconvert/exporters/qt_screenshot.py

```diff
@@ -0,0 +1,51 @@
+from PyQt5.QtWidgets import QApplication
+from PyQt5 import QtCore
+from PyQt5.QtWebEngineWidgets import QWebEngineView, QWebEngineSettings
+from PyQt5.QtGui import QPageLayout, QPageSize
+
+
+class QtScreenshot(QWebEngineView):
+
+    def __init__(self, app):
+        super().__init__()
+        self.app = app
+
+    def capture(self, url, output_file, paginate):
+        self.output_file = output_file
+        self.paginate = paginate
+        self.load(QtCore.QUrl(url))
+        self.loadFinished.connect(self.on_loaded)
+        # Create hidden view without scrollbars
+        self.setAttribute(QtCore.Qt.WA_DontShowOnScreen)
+        self.page().settings().setAttribute(
+            QWebEngineSettings.ShowScrollBars, False)
+        if output_file.endswith(".pdf"):
+            self.export = self.export_pdf
+            self.page().pdfPrintingFinished.connect(lambda *args: self.app.exit())
+        elif output_file.endswith(".png"):
+            self.export = self.export_png
+        else:
+            raise RuntimeError(f"Export file extension not supported: {output_file}")
+        self.show()
+        self.app.exec()
+
+    def on_loaded(self):
+        self.size = self.page().contentsSize().toSize()
+        self.resize(self.size)
+        # Wait for resize
+        QtCore.QTimer.singleShot(1000, self.export)
+
+    def export_pdf(self):
+        if self.paginate:
+            page_size = QPageSize(QPageSize.A4)
+            page_layout = QPageLayout(page_size, QPageLayout.Portrait, QtCore.QMarginsF())
+        else:
+            factor = 0.75
+            page_size = QPageSize(QtCore.QSizeF(self.size.width() * factor, self.size.height() * factor), QPageSize.Point)
```

Comment:
`self.page().contentsSize()` already returns a `QSizeF` object while `QSizeF.toSize` returns a `QSize` object.

We should be able to get away with

```python
page_size = QPageSize(self.page().contentsSize(), QPageSize.Point)
```

Reply:
then we should be able to drop the line `self.size = self.page().contentsSize().toSize()`.

Reply:
Hum, actually, looking at the unit ` QPageSize.Point`, it looks a bit arbitrary. I think we should determine that based on the DPI ratio of the "device".

Reply:
https://lists.qt-project.org/pipermail/qtwebengine/2017-May/000453.html

Reply:
> ```python
> page_size = QPageSize(self.page().contentsSize(), QPageSize.Point)
> ```

I had already tried that, but it generates a PDF with a larger size than the required page size.

> Hum, actually, looking at the unit ` QPageSize.Point`, it looks a bit arbitrary. I think we should determine that based on the DPI ratio of the "device".

I agree, I've been through these threads, but I couldn't find any relevant information about the DPI ratio.

Reply:
Maybe our friends at the @qt company would know how to achieve that in a clean fashion. cc @cmaureir

Reply:
uh...I really don't have an answer, and I would expect `QPageSize.Point` to get the DPI scaling, but if that's not the case...then there might be a better way. I will ask to devs that work more actively on QtPdf

Reply:
So, I asked around and it seems something similar was apply here: https://codereview.qt-project.org/c/qt/qtwebengine/+/277187 specifically:
```
QRectF pageRect = m_printer->pageRect(QPrinter::DevicePixel);
documentSize = documentSize.scaled(pageRect.size(), Qt::KeepAspectRatio);
```

Sorry for not having a direct answer to the issue, but if you had a case where `QPageSize.Point` has a strange behavior, it would be better to communicate with the devs directly through a report https://bugreports.qt.io/projects/QTBUG/ using the component `PDF`. Let me know if I can somehow help with something else!

Reply:
I have been doing some more experiments with this and I must say that I am very confused. I am enclined to merge the pull request as it is.

Reply:
@cmaureir I have spent some time today trying to figure this one out. I really think that the behavior of PyQtWebEngine here is buggy. Would love to talk about this in greater depth to our @qt friends.

Reply:
As long as it's a general Qt issue, and not PyQt specific I can try to get your question to the people working in QtPdf :+1:  but if it's PyQt specific, you will need to reach the folks at Riverbank.

Reply:
I think it is likely to be Qt. I will post something a bit more specific.

Reply:
@cmaureir the bug is that printToPdf adds whitespace at the end of the document. It does not do so when exporting to PNG. (This only occurs when we set pagination to False or course). Multiplicating the dimensions by 0.75 magically makes all whitespace disapear.

![Screenshot from 2022-01-26 14-10-08](https://user-images.githubusercontent.com/2397974/151169359-b28f5d40-0976-40e3-a4c6-376a73439e51.png)